SerialPrograms/Source/PokemonSV/Programs/ItemPrinter/PokemonSV_ItemPrinterRNG.cpp

```diff
@@ -629,6 +704,92 @@ void ItemPrinterRNG::run_item_printer_rng(
+){
+    uint32_t num_happiny_dust = check_num_happiny_dust(env, context);
+
+    // give a buffer of 50, for a margin of safety. signed int to handle negative numbers
+    int32_t num_happiny_dust_can_use = num_happiny_dust - MIN_HAPPINY_DUST - 50; 
+    num_happiny_dust_can_use = num_happiny_dust_can_use < 0 ? 0 : num_happiny_dust_can_use;
+
+    // assume 62% value for Happiny Dust to account for item printer wasteage.
+    uint32_t num_print_jobs = num_happiny_dust_can_use * 0.62; // truncate the float back to int
+    env.console.log("Number of Happiny Dust we have: " + std::to_string(num_happiny_dust));
+    env.console.log("Number of Happiny Dust we can use (with some safety margins): " + std::to_string(num_happiny_dust_can_use));
+    env.console.log("Number of print jobs we can do before material farming: " + std::to_string(num_print_jobs));
+    return num_print_jobs;
+}
+
+uint32_t ItemPrinterRNG::check_num_happiny_dust(
+    SingleSwitchProgramEnvironment& env, BotBaseContext& context
+){
+    ItemPrinterRNG_Descriptor::Stats& stats = env.current_stats<ItemPrinterRNG_Descriptor::Stats>();
+    env.log("Check how much Happiny Dust we have.");
+    uint32_t num_happiny_dust;
+    while (true){
+        context.wait_for_all_requests();
+
+        OverworldWatcher overworld(COLOR_BLUE);
+        AdvanceDialogWatcher dialog(COLOR_RED);
+        PromptDialogWatcher prompt(COLOR_GREEN);
+        DateChangeWatcher date_reader;
+        WhiteButtonWatcher material(COLOR_GREEN, WhiteButton::ButtonX, {0.63, 0.93, 0.17, 0.06});
+        int ret = wait_until(
+            env.console, context, std::chrono::seconds(120),
+            {
+                overworld,
+                dialog,
+                prompt,
+                material,
+            }
+        );
+        switch (ret){
+        case 0:
+            env.log("Detected overworld... Entering item printer.");
+            pbf_press_button(context, BUTTON_A, 20, 30);
+            continue;
+
+        case 1:
+            env.log("Detected advance dialog.");
+            pbf_press_button(context, BUTTON_B, 20, 30);
+            continue;
+
+        case 2:{
+            env.log("Detected prompt dialog. Entering item printer.");
+            pbf_press_button(context, BUTTON_A, 10, 30);
+            context.wait_for_all_requests();
+            continue;
+        }
+        case 3:{
+            env.log("Detected material selection.");
+            ItemPrinterMaterialDetector detector(COLOR_RED);
+            // Press the right button 6 times to see the Happiny Dust in the material selector.
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);
+            pbf_press_dpad(context, DPAD_RIGHT, 20, 30);

```

Comment:
What happens if Happiny dust is not visible after 6 scrolls? Say you're missing most of the earlier ones so that 6 scrolls past it.

Reply:
True. I forgot about this. 
Number recognition is faster and more reliable right?
I'm thinking we use number recognition to find materials with 68% value, of which there are only 4 materials according to Serebii (Ditto Goo, Happiny Dust, Magby hair, Beldum Claw). Then use string recognition.

Reply:
Number is worse because there's no error correction. Strings have error correction since we can compare against all the possible string values. And there's more characters to work with.